rustc_codegen_ssa: add time-passes output for each CGU compilation

tgnottingham opened the pull request:
Add time-passes event for each CGU compilation to LLVM IR. This is
useful to measure, as the event can cause a large increase in RSS.

Also, allow `extra_verbose_generic_activity` to accept multiple event
arguments.

rust-highfive commented:
r? @matthewjasper

(rust-highfive has picked a reviewer for you, use r? to override)

tgnottingham commented:
In conjunction with #81536, sample of `-Z time-passes` output for `rustc_middle`:

```
time:   0.000; rss:  1756MB ->  1756MB	find_cgu_reuse
time:   2.810; rss:  1756MB ->  1921MB	compile_codegen_unit(47luyf1dq9ttmk2y)
time:   3.587; rss:  1921MB ->  2204MB	compile_codegen_unit(5banyx1ghjvcxyne)
time:   4.355; rss:  2204MB ->  2482MB	compile_codegen_unit(2goq7lwe3l6np96t)
time:   3.161; rss:  2482MB ->  2712MB	compile_codegen_unit(qpagbtzbgw0u8oj)
time:   3.477; rss:  2712MB ->  2948MB	compile_codegen_unit(3mkkqle0ppntfa3i)
time:   1.266; rss:  2948MB ->  3057MB	compile_codegen_unit(47cjpiuhvj52boxv)
time:   0.680; rss:  3057MB ->  3114MB	compile_codegen_unit(4nihpc2q9xthln9l)
time:   1.846; rss:  3114MB ->  3261MB	compile_codegen_unit(z93ybhzpgfz689e)
time:   0.667; rss:  3261MB ->  3300MB  LLVM_module_optimize_function_passes(2goq7lwe3l6np96t)
```

See also https://github.com/rust-lang/rust/pull/81536#issuecomment-770047544.

bjorn3 commented:
It would be nice to also add it to https://github.com/rust-lang/rust/blob/fee0d31397c5ac46d08867e903131d1d73825a9e/compiler/rustc_codegen_cranelift/src/driver/aot.rs#L120

rust-log-analyzer commented:
The job **`mingw-check`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/runs/1835464547) [(plain)](https://github.com/rust-lang/rust/commit/d5ee6978667e36374473dbe8741e021e2860e676/checks/1835464547/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
    Checking rustdoc v0.0.0 (/checkout/src/librustdoc)
error[E0308]: mismatched types
  --> src/librustdoc/formats/renderer.rs:69:60
   |
69 |         .extra_verbose_generic_activity("create_renderer", T::descr())
   |                                                            ^^^^^^^^^^ expected slice, found `str`
   |
   = note: expected reference `&[_]`
              found reference `&'static str`
error[E0308]: mismatched types
   --> src/librustdoc/formats/renderer.rs:109:65
    |
    |
109 |     prof.extra_verbose_generic_activity("renderer_after_krate", T::descr())
    |                                                                 ^^^^^^^^^^ expected slice, found `str`
    |
    = note: expected reference `&[_]`
               found reference `&'static str`
error: aborting due to 2 previous errors

For more information about this error, try `rustc --explain E0308`.
error: could not compile `rustdoc`
```

</details>

tgnottingham commented:
Changed things a bit, but the end result is basically the same. Instead of adding a `compile_codegen_unit` event, I upgraded the `codegen_module` profile event to be included in `-Z time-passes` output, and added the same event for the cranelift backend.

JohnCSimon commented:
@rustbot label: -S-waiting-on-review +S-waiting-on-author

crlf0710 commented:
@tgnottingham Ping from triage! What's the current status of this?

tgnottingham commented:
@crlf0710 Closing this, as I don't have the bandwidth right now and it's not of high importance.
